org.eclipse.xtext/src/org/eclipse/xtext/resource/impl/ChunkedResourceDescriptions.java

```diff
@@ -0,0 +1,207 @@
+	protected ResourceSet resourceSet;
+
+	public ChunkedResourceDescriptions() {
+	}
+
+	public ChunkedResourceDescriptions(Map<String, ResourceDescriptionsData> initialData) {
+		chunk2resourceDescriptions = new ConcurrentHashMap<>(initialData);
+	}
+
+	public ChunkedResourceDescriptions(Map<String, ResourceDescriptionsData> initialData, ResourceSet resourceSet) {
+		this(initialData);
+		setResourceSet(resourceSet);
+	}
+
+	/**
+	 * Creates a shallow copy of the resource descriptions map and installs it with the given ResourceSet.
+	 */
+	public ChunkedResourceDescriptions createShallowCopyWith(ResourceSet resourceSet) {
+		return new ChunkedResourceDescriptions(chunk2resourceDescriptions, resourceSet);
+	}
+
+	public ResourceSet getResourceSet() {
+		return this.resourceSet;
+	}
+
+	protected void setResourceSet(ResourceSet resourceSet) {
+		if (this.resourceSet != null)
+			throw new IllegalStateException(
+					"This " + getClass().getName() + " is already associated with a different resource set.");
+		if (ChunkedResourceDescriptions.findInEmfObject(resourceSet) != null)
+			throw new IllegalStateException(
+					"There is already a different " + getClass().getName() + " installed in the given resource set.");
+		this.resourceSet = resourceSet;
+		attachToEmfObject(resourceSet);
+	}
+
+	public ResourceDescriptionsData setContainer(String name, ResourceDescriptionsData descriptions) {
+		return chunk2resourceDescriptions.put(name, descriptions);
+	}
+
+	public ResourceDescriptionsData removeContainer(String name) {
+		return chunk2resourceDescriptions.remove(name);
+	}
+
+	@Override
+	protected Iterable<? extends ISelectable> getSelectables() {
+		return chunk2resourceDescriptions.values();
+	}
+
+	@Override
+	public Iterable<IResourceDescription> getAllResourceDescriptions() {
+		return concat(map(chunk2resourceDescriptions.values(), it -> it.getAllResourceDescriptions()));
+	}
+
+	@Override
+	public IResourceDescription getResourceDescription(URI uri) {
+		for (ResourceDescriptionsData selectable : chunk2resourceDescriptions.values()) {
+			IResourceDescription result = selectable.getResourceDescription(uri);
+			if (result != null)
+				return result;
+		}
+		return null;
+	}
+
+	public ResourceDescriptionsData getContainer(URI uri) {
+		for (ResourceDescriptionsData container : chunk2resourceDescriptions.values()) {
+			IResourceDescription description = container.getResourceDescription(uri);
+			if (description != null)
+				return container;
+		}
+		return null;
+	}
+
+	public ResourceDescriptionsData getContainer(String containerHandle) {
+		return chunk2resourceDescriptions.get(containerHandle);
+	}
+
+	@Override
+	@SuppressWarnings("unused")
+	public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+		int numChunks = in.readInt();
+		for (Integer i : new ExclusiveRange(0, numChunks, true)) {
```

Comment:
a plain forLoop `for (int i = 0; i < numChunks; i++) ` would be better.

Reply:
Will do.